src/backend/distributed/connection/connection_management.c

```diff
@@ -694,8 +694,27 @@ StartConnectionEstablishment(ConnectionHashKey *key)
 	entry = hash_search(ConnParamsHash, key, HASH_ENTER, &found);
 	if (!found || !entry->isValid)
 	{
-		/* if they're not found, compute them from GUC, runtime, etc. */
-		GetConnParams(key, &entry->keywords, &entry->values, ConnectionContext);
+		if (found && !entry->isValid)
+		{
+			char **keyword = &entry->keywords[entry->nonGlobalParamStart];
```

Comment:
But if we start from `nonGlobalParamStart`, wouldn't we hit the issue that Marco raised here: https://github.com/citusdata/citus/pull/2611/files#r259832277

On any invalidation, we'd re-allocate items that are `nonGlobalParams` in the `ConnectionContext`.

I had an example case to repro in that thread, which seem to have the same memory leak affect on this branch:
```SQL
-- session 1
SELECT count(*) FROM table;
\watch 0.001
```
and:

```SQL
-- session 2
alter system set citus.node_conninfo TO 'connect_timeout=20';
SELECT pg_reload_conf();
\watch 0.001 
```

(Well, I've created a script with different values for connect timeout and used it, so not sure about setting the same value always works as I expect ):
```
SELECT pg_reload_conf();
alter system set citus.node_conninfo TO 'connect_timeout= 6 ';
SELECT pg_reload_conf();
alter system set citus.node_conninfo TO 'connect_timeout= 7 ';
SELECT pg_reload_conf();
alter system set citus.node_conninfo TO 'connect_timeout= 8 ';
SELECT pg_reload_conf();
alter system set citus.node_conninfo TO 'connect_timeout= 9 ';
SELECT pg_reload_conf();
....
```

Reply:
I don't quite understand the issue. It's odd that there's still a memory leak; I _didn't_ change my code to allocate the global settings in the `ConnectionContext`; they're still managed by the configuration methods, and `malloc`ed/`free`d there, no?

Reply:
Oh dur, I see the problem now; the arrays aren't `pfree`d. I'll see whether that fixes your memory issue.

Reply:
Ok, so @onderkalaci… I think I've removed any leaks in _this_ code, but the _side effects_ of this code may have some leaks? Like, shouldn't we somehow tear down the existing cached connections if their parameters change?

I'm starting to lean towards "maybe this GUC should require a restart, or at least a new backend"?

Reply:
> Like, shouldn't we somehow tear down the existing cached connections if their parameters change?

I think yes, that'd be ideal. But given those are too rare, the assumption in the code looks like `we can live with that`.

(But the memory leak might not be too rare because any SIGHUP could trigger that)

> I'm starting to lean towards "maybe this GUC should require a restart, or at least a new backend"?

I think that used to be case, but we recently had to change it to enable SSL by default. See the related change [here](https://github.com/citusdata/citus/commit/4af40eee76d3e17260e0a4af8febb18039d6c9eb).

As far as I can see, that's only required for upgrades from non-ssl-enabled clusters to later versions of Citus where ssl-enabled.

Is there a way to enable SIGHUP only during the upgrades?